optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,134 @@ private void initBenchmarkDirectoryAndSubdirs() {
-            List<ProblemBenchmarkResult> unifiedProblemBenchmarkResultList = plannerBenchmarkResult.getUnifiedProblemBenchmarkResultList();
-            Iterator<ProblemBenchmarkResult> it = unifiedProblemBenchmarkResultList.iterator();
-            while (timeLeft > 0L) {
-                if (!it.hasNext()) {
-                    it = unifiedProblemBenchmarkResultList.iterator();
-                }
-                ProblemBenchmarkResult problemBenchmarkResult = it.next();
-                timeLeft = warmUp(problemBenchmarkResult, startingTimeMillis, plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit(), timeLeft);
+        if (plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit() <= 0L) {
+            return;
+        }
+        logger.info("================================================================================");
+        logger.info("Warm up started");
+        logger.info("================================================================================");
+        long timeLeftTotal = plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit();
+        int parallelBenchmarkCount = plannerBenchmarkResult.getParallelBenchmarkCount();
+        int solverBenchmarkResultCount = plannerBenchmarkResult.getSolverBenchmarkResultList().size();
+        long cyclesCount = Math.round(Math.ceil(solverBenchmarkResultCount / (double) parallelBenchmarkCount));
+        long timeLeftPerCycle = Math.round(Math.floor((timeLeftTotal / (double) cyclesCount)));
+        ConcurrentMap<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner> futureMap
+                = new ConcurrentHashMap<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner>(parallelBenchmarkCount);
+        Map<SolverBenchmarkResult, TerminationConfig> originalTerminationConfigMap
+                = new HashMap<SolverBenchmarkResult, TerminationConfig>(solverBenchmarkResultCount);
+        ConcurrentMap<SolverBenchmarkResult, Integer> singleBenchmarkResultIndexMap
+                = new ConcurrentHashMap<SolverBenchmarkResult, Integer>(solverBenchmarkResultCount);
+
+        warmUpStarted(originalTerminationConfigMap);
+        SolverBenchmarkResult[] solverBenchmarkResultCycleArray = new SolverBenchmarkResult[parallelBenchmarkCount];
+        int solverBenchmarkResultIndex = 0;
+        for (int i = 0; i < cyclesCount; i++) {
+            long timeCycleEnd = System.currentTimeMillis() + timeLeftPerCycle;
+            for (int j = 0; j < parallelBenchmarkCount; j++) {
+                solverBenchmarkResultCycleArray[j] = plannerBenchmarkResult.
+                        getSolverBenchmarkResultList().get(solverBenchmarkResultIndex % solverBenchmarkResultCount);
+                solverBenchmarkResultIndex++;
+            }
+            warmUpPopulate(futureMap, singleBenchmarkResultIndexMap, solverBenchmarkResultCycleArray, timeLeftPerCycle);
+            warmUp(futureMap, singleBenchmarkResultIndexMap, timeCycleEnd);
+            futureMap.clear();
```

Comment:
Wouldn't it be more memory friendly if it is just reused and not allocated/deallocated each time?